### ISSUE TITLE
GLSL C++ constexpr Compiler

### ISSUE THREAD

**Issue by alaingalvan** [KhronosGroup/glslang#623]
Recently on [Hacker News](https://news.ycombinator.com/item?id=12988696) there was a project that took in a literal string `""` of C code and compiled it called [constexpr-8cc](https://github.com/kw-udon/constexpr-8cc). 

so I'm wondering if it would be possible to do the same for GLSL or HLSL with glslang. I think this would make developing Vulkan/OpenGL 4.6 apps much easier as a result, but this does come with some bad sides:

- Your program would consistently take longer to compile (unless this could somehow take advantage of caching).
- Shaders would be embeded in the executable, resulting in a larger sized output. (However this may be good for people trying to obfuscate their shaders.)

```cpp
#include "vulkan.hpp"
#include "spirv-compiler.h"

int main()
{
  //  ...

  // After setting up a lot of Vulkan Constructs
  constexpr buffer buf = 
spirv::compile(
"#version 450"
"#extension GL_ARB_separate_shader_objects : enable"
"#extension GL_ARB_shading_language_420pack : enable"
"layout (location = 0) in vec3 inColor;"
"layout (location = 0) out vec4 outFragColor;"
"void main() "
"{"
"  outFragColor = vec4(inColor, 1.0);"
"}"
);
  constexpr uint32_t output_size = buf.size;

  auto fragModule = device.createShaderModule(
		vk::ShaderModuleCreateInfo(
			vk::ShaderModuleCreateFlags(),
			output_size,
			(uint32_t*) buf.b
		)
	);

  // ...
}
```

**Comment by Corillian** [comment]
My 2 cents is that while this is pretty high on the novelty scale it's pretty low on the practicality scale. Having all of your shaders embedded as strings in source code would be a complete maintenance nightmare in all but the simplest of applications. That's also not even considering the fact that any reasonably sized project likely has tools that allow artists to play around with creating materials which of course is entirely dependent on dynamically recompiling shaders.

To summarize I think effort would be much better spent elsewhere.
